kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopCredentialsManager.scala

```diff
@@ -0,0 +1,225 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.credentials
+
+import java.time.Duration
+import java.util.ServiceLoader
+import java.util.concurrent._
+
+import scala.collection.mutable
+import scala.util.{Failure, Success, Try}
+
+import com.google.common.cache.{CacheBuilder, CacheLoader, LoadingCache}
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.security.Credentials
+
+import org.apache.kyuubi.{KyuubiException, Logging}
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.config.KyuubiConf._
+import org.apache.kyuubi.service.AbstractService
+import org.apache.kyuubi.util.{KyuubiHadoopUtils, ThreadUtils}
+
+class HadoopCredentialsManager private (name: String) extends AbstractService(name)
+    with Logging {
+
+  def this() = this(classOf[HadoopCredentialsManager].getSimpleName)
+
+  private val userRefMap = new ConcurrentHashMap[String, UserCredentialsRef]()
+
+  private var engineRefCache: LoadingCache[String, EngineCredentialsRef] = _
+
+  private var providers: Map[String, HadoopDelegationTokenProvider] = _
+  private var requiredProviders: Map[String, HadoopDelegationTokenProvider] = _
+
+  private var renewalInterval: Long = _
+  private var renewalRetryWait: Long = _
+  private var renewalExecutor: ScheduledExecutorService = _
+  private var hadoopConf: Configuration = _
+
+  override def initialize(conf: KyuubiConf): Unit = {
+    val cacheLoader = new CacheLoader[String, EngineCredentialsRef] {
+      override def load(key: String): EngineCredentialsRef = new EngineCredentialsRef
+    }
+    engineRefCache = CacheBuilder.newBuilder()
+      .concurrencyLevel(conf.get(KyuubiConf.SERVER_EXEC_POOL_SIZE))
```

Comment:
why we need SERVER_EXEC_POOL_SIZE  here?

Reply:
Thought to set  a proper concurrency level for `engineRefCache`. 
But since loading and getting from `engineRefCache` are almost an instant action and won't become a bottleneck, maybe it is better to remove it.